Source/WebKit/UIProcess/ProvisionalPageProxy.cpp

```diff
@@ -202,15 +203,30 @@ void ProvisionalPageProxy::initializeWebPage(RefPtr<API::WebsitePolicies>&& webs
             parameters.openerFrameIdentifier = openerFrame->frameID();
             parameters.mainFrameIdentifier = m_page->mainFrame()->frameID();
 
-            auto remotePageProxy = RemotePageProxy::create(*openerPage, m_process, navigationDomain);
-            remotePageProxy->injectPageIntoNewProcess();
-            openerPage->addOpenedRemotePageProxy(WTFMove(remotePageProxy));
+            if (auto existingRemotePageProxy = m_page->takeRemotePageProxyInOpenerProcessIfDomainEquals(navigationDomain)) {
+                ASSERT(existingRemotePageProxy->process().processID() == m_process->processID());
+                m_webPageID = existingRemotePageProxy->pageID();
+                m_mainFrame = existingRemotePageProxy->page()->mainFrame();
+                m_messageReceiverRegistration.stopReceivingMessages();
+                m_messageReceiverRegistration.transferMessageReceivingFrom(existingRemotePageProxy->messageReceiverRegistration(), *this);
+                LocalFrameCreationParameters localFrameCreationParameters {
+                    std::nullopt
+                };
+                m_process->send(Messages::WebPage::TransitionFrameToLocal(localFrameCreationParameters, m_page->mainFrame()->frameID()), m_webPageID);
+                sendPageCreationParameters = false;
+            } else {
+                auto remotePageProxy = RemotePageProxy::create(*openerPage, m_process, navigationDomain);
+                remotePageProxy->injectPageIntoNewProcess();
```

Comment:
Won't we send CreateWebPage both here and below (on 227) when entering this else?

Reply:
Keeping track of all the pages and processes is tricky!